### PR TITLE
Feature/recover user posts

### PR DESCRIPTION
**Pull request by angel-chaves**
This pull request introduces a new feature for fetching paginated user posts via an authenticated API endpoint. It includes changes to the backend logic, database queries, API routes, and test cases. Below is a summary of the most important changes grouped by theme:

This pull request introduces a new feature for managing user posts via a REST API, along with enhancements to the project's configuration and documentation. 

# THE MOST IMPORTANT TO KNOW ABOUT:
* Implements Pagination to get user posts that are visible.
* See [Post API Description](https://app.clickup.com/9013818188/v/dc/8cm7wuc-873/8cm7wuc-833) For an example of how works the endpoint.
* Read from database the posts, there are testing posts for angel.chaveschinchilla@ucr.ac.cr user
* There are A LOT of files because there are a lot of testing.

### New Feature: Fetch Paginated User Posts
* Added a `GET /api/user/posts/mine` endpoint to fetch paginated posts for the authenticated user. The endpoint validates query parameters, checks user authorization, and retrieves posts from the database. (`src/features/posts/routes/post.routes.ts`, `src/features/posts/controllers/getPosts.controller.ts`, `src/features/posts/services/getPosts.service.ts`) [[1]](diffhunk://#diff-16ea406c400345b52dc2700356e2a5808601577c9ddf8469cc8c6b15d829c393R3-R10) [[2]](diffhunk://#diff-b5dfcfed527495433f70b8845ddb22388ee17d591242571bfe3eb8388053715aR1-R43) [[3]](diffhunk://#diff-a8b55df1ade70f74ca70194933e4b57beb326b7d5bf109d06648c8be38b8a4b3R1-R45)

### Database Integration
* Implemented repository functions for fetching paginated posts (`getVisiblePostsByUserIdPaginated`) and counting total posts (`getTotalVisiblePostsByUserId`) from the database. These functions include SQL queries with pagination and filtering logic. (`src/features/posts/repositories/getPosts.repository.ts`)

### Data Validation and DTOs
* Introduced `getUserPostsSchema` for validating query parameters (e.g., `page` and `limit`) using Yup. Added a `GetUserPostsDTO` type for strongly-typed data transfer. (`src/features/posts/dto/getUserPosts.dto.ts`)

### Testing
* Added integration tests for the new controller to validate behavior for authenticated and unauthenticated users, as well as invalid query parameters. (`test-api/controllers/userPosts.controller.test.ts`)
* Added repository tests for verifying database query behavior and edge cases (e.g., no posts found). (`test-api/repositories/user.posts.repository.test.ts`)
* Added route-level tests to ensure the endpoint functions correctly with mocked middleware and controller. (`test-api/routes/getPosts.routes.test.ts`)

### Environment Configuration
* Added test-specific environment variables for database and Firebase configuration in `.env.test`. (`.env.test`)

### CODE REVIEW

File: src/features/posts/services/userPosts.service.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    const totalPosts = await getTotalVisiblePostsByEmail(email);
+    const totalPages = Math.ceil(totalPosts / limit);
+
+    const posts = await getVisiblePostsByEmailPaginated(email, offset, limit);
+
+    if (!posts) {
+      throw new InternalServerError('Failed to fetch posts');
+    }
+
+    console.log('Posts:', posts);
```
Comment (CristopherH11) [inline src/features/posts/services/userPosts.service.ts:26]: Eliminar el log para producción, o limitarlo a información mas resumida (posts.id)

File: src/app.ts [hunk 2]
```diff
@@ -11,6 +12,9 @@ app.get('/', (req, res) => {
 app.use(express.json());
 app.use(cors());
 
+// Add the user posts routes
+app.use('/api', userPostsRouter);
```
Comment (CristopherH11) [inline src/app.ts:16]: Se sugiere que sea una ruta general
```suggestion
app.use('/api/posts', postsRouter);
```

File: src/features/posts/routes/userPosts.routes.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+import { Router, RequestHandler } from 'express';
+import { getUserPostsController } from '../controllers/userPosts.controller';
+import { authenticateJWT } from '../../middleware/authenticate.middleware';
+
+const router = Router();
+
+router.get('/posts/mine', authenticateJWT, getUserPostsController as RequestHandler);
```
Comment (CristopherH11) [inline src/features/posts/routes/userPosts.routes.ts:7]:
```suggestion
router.get('/mine', authenticateJWT, getUserPostsController as RequestHandler);
```

**Comment by SrDavidC** [comment]
Added suggestions review comments by @CristopherH11and @SrDavidC

File: src/features/posts/interfaces/posts.entities.interface.ts [hunk 4]
```diff
@@ -0,0 +1,14 @@
+export interface BasePost {
+  id: string;
```
Comment (CristopherH11) [inline src/features/posts/interfaces/posts.entities.interface.ts:2]: Verificar con la base de datos que si es igual

**Review by CristopherH11 — APPROVED** [review]

**Review by SrDavidC — APPROVED** [review]
Check Cris' suggestions.

**Review by johanawn — APPROVED** [review]

File: src/features/posts/controllers/userPosts.controller.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+export const getUserPostsController = async (req: AuthenticatedRequest, res: Response, next: NextFunction) => {
+  try {
+    // Validate and cast the query parameters to GetUserPostsDTO
+    const validatedQuery = await getUserPostsSchema.validate(req.query, {
+      abortEarly: false,
+      stripUnknown: true,
+    }) as GetUserPostsDTO;
+
+    // Check if the user is authenticated
```
[the remark this answers is not in the document]
Comment (angel-chaves) [reply src/features/posts/controllers/userPosts.controller.ts:16]: Just in english please, i don't understand you.

File: src/features/posts/interfaces/posts.entities.interface.ts (hunk 4 above)
Comment (angel-chaves) [reply src/features/posts/interfaces/posts.entities.interface.ts:2]: Done

Comment (angel-chaves) [reply src/features/posts/interfaces/posts.entities.interface.ts:2]: Some of other suggestions will be taken in account in futures PRs related to refactor

File: src/features/posts/services/userPosts.service.ts (hunk 1 above)
Comment (angel-chaves) [reply src/features/posts/services/userPosts.service.ts]: Done

File: src/features/posts/controllers/userPosts.controller.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (angel-chaves) [reply src/features/posts/controllers/userPosts.controller.ts]: Done
